test/AllocatorRedeemer.t.sol

```diff
@@ -0,0 +1,183 @@
+
+        vm.expectRevert("AllocatorRedeemer/system-live");
+        redeemer.pull(address(usdc));
+
+        assertEq(vat.live(), 1);
+        vat.cage();
+        assertEq(vat.live(), 0);
+
+        assertEq(nst.balanceOf(address(buffer)), 10_000_000 * 10**18);
+        assertEq(nst.balanceOf(address(redeemer)), 0);
+        vm.expectEmit(true, true, true, true);
+        emit Pull(address(nst), 10_000_000 * 10**18);
+        redeemer.pull(address(nst));
+        assertEq(nst.balanceOf(address(buffer)), 0);
+        assertEq(nst.balanceOf(address(redeemer)), 10_000_000 * 10**18);
+
+        assertEq(usdc.balanceOf(address(buffer)), 10_000_000 * 10**6);
+        assertEq(usdc.balanceOf(address(redeemer)), 0);
+        vm.expectEmit(true, true, true, true);
+        emit Pull(address(usdc), 10_000_000 * 10**6);
+        redeemer.pull(address(usdc));
+        assertEq(usdc.balanceOf(address(buffer)), 0);
+        assertEq(usdc.balanceOf(address(redeemer)), 10_000_000 * 10**6);
+
+        // Simulate skim (only 40% of the totalSupply of the token)
+        vat.grab(ilk, address(vault), address(this), address(0), -400_000 * 10**18, -int256(art));
+
+        gemJoin.exit(user1, 100_000 * 10**18);
+        gemJoin.exit(user2, 100_000 * 10**18);
+        gemJoin.exit(user3, 100_000 * 10**18);
+        gemJoin.exit(user4, 100_000 * 10**18);
+
+        vm.prank(user1); gem.approve(address(redeemer), type(uint256).max);
+        vm.prank(user2); gem.approve(address(redeemer), type(uint256).max);
+        vm.prank(user3); gem.approve(address(redeemer), type(uint256).max);
+        vm.prank(user4); gem.approve(address(redeemer), type(uint256).max);
+
+        vm.expectRevert("AllocatorRedeemer/wad-zero");
+        vm.prank(user1); redeemer.pack(0);
+        vm.prank(user1); redeemer.pack(80_000 * 10**18);
+        assertEq(redeemer.bag(user1), 80_000 * 10**18);
+        vm.prank(user1); redeemer.pack(10_000 * 10**18);
+        assertEq(redeemer.bag(user1), 90_000 * 10**18);
+        vm.prank(user1); redeemer.pack(10_000 * 10**18);
+        assertEq(redeemer.bag(user1), 100_000 * 10**18);
+        vm.expectRevert("Gem/insufficient-balance");
+        vm.prank(user1); redeemer.pack(1);
+        vm.prank(user2); redeemer.pack(100_000 * 10**18);
+        vm.prank(user3); redeemer.pack(100_000 * 10**18);
+        vm.prank(user4); redeemer.pack(100_000 * 10**18);
+
+        vm.expectRevert("AllocatorRedeemer/wad-zero");
+        vm.prank(user1); redeemer.cash(address(nst), 0);
+        vm.prank(user1); redeemer.cash(address(nst), 50_000 * 10**18);
+        assertEq(nst.balanceOf(user1), 1_250_000 * 10**18);
+        assertEq(redeemer.out(address(nst), user1), 50_000 * 10**18);
+        assertEq(redeemer.totOut(address(nst)), 50_000 * 10**18);
+        vm.prank(user1); redeemer.cash(address(nst), 50_000 * 10**18);
+        assertEq(nst.balanceOf(user1), 2_500_000 * 10**18);
+        assertEq(nst.balanceOf(address(redeemer)), 7_500_000 * 10**18);
```

Comment:
Seems like `nst.balanceOf(address(redeemer)` is only checked after some of the operations.

Reply:
Did a quick check and can't find where is missing.

Reply:
look after the previous redeemer.cash call.